pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+			Ok(())
+		}
+
+		#[pallet::weight(100)]
+		pub fn order_supply(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranche_id: T::TrancheId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// TODO: Ensure this account is authorized for this tranche
+			let (currency, epoch) = {
+				let pool = Pool::<T>::try_get(pool_id).map_err(|_| Error::<T>::NoSuchPool)?;
+				ensure!(pool.closing_epoch.is_none(), Error::<T>::PoolClosing);
+				(pool.currency, pool.current_epoch)
+			};
+			let tranche = TrancheLocator {
+				pool_id,
+				tranche_id,
+			};
+			let pool_account = PoolLocator { pool_id }.into_account();
+
+			Order::<T>::try_mutate(&tranche, &who, |order| -> DispatchResult {
+				if amount > order.supply {
+					let transfer_amount = amount - order.supply;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_supply = &mut pool.tranches[tranche_id.into()].epoch_supply;
+						*epoch_supply = epoch_supply
+							.checked_add(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &who, &pool_account, transfer_amount)
+					})?;
+				} else if amount < order.supply {
+					let transfer_amount = order.supply - amount;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_supply = &mut pool.tranches[tranche_id.into()].epoch_supply;
+						*epoch_supply = epoch_supply
+							.checked_sub(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &pool_account, &who, transfer_amount)
+					})?;
+				}
+				order.supply = amount;
+				order.epoch = epoch;
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(100)]
+		pub fn order_redeem(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranche_id: T::TrancheId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// TODO: Ensure this account is authorized for this tranche
+			let epoch = {
+				let pool = Pool::<T>::try_get(pool_id).map_err(|_| Error::<T>::NoSuchPool)?;
+				ensure!(pool.closing_epoch.is_none(), Error::<T>::PoolClosing);
+				pool.current_epoch
+			};
+			let currency = T::TrancheToken::tranche_token(pool_id, tranche_id);
+			let tranche = TrancheLocator {
+				pool_id,
+				tranche_id,
+			};
+			let pool_account = PoolLocator { pool_id }.into_account();
+
+			Order::<T>::try_mutate(&tranche, &who, |order| -> DispatchResult {
+				if amount > order.redeem {
+					let transfer_amount = amount - order.redeem;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_redeem = &mut pool.tranches[tranche_id.into()].epoch_redeem;
+						*epoch_redeem = epoch_redeem
+							.checked_add(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &who, &pool_account, transfer_amount)
```

Comment:
Can this fail? Mustn't we have an early return here too? Otherwise, `T::Tokens::transfer` might fail, but we still insert `Order`?

Reply:
If this fails the entire `try_mutate` is aborted, so no changes are written